[Complex-Dom] Combine visibility: hidden and display: none for side TreeView hidden elements.

lpardosixtosMs opened the pull request:
Following the discussion on #316 this PR adds a new class name to TreeView elements that should be hidden using `display: none`.

After this change the sidebar has 6004 elements and 2996 of them had the `display: none` value or are descendants of elements with `display: none` value. 

Hosted version: https://lpardosixtosms.github.io/SpeedometerHiddenNone/#home

lpardosixtosMs commented:
### Before and after results

<img width="757" alt="image" src="https://github.com/WebKit/Speedometer/assets/94007365/7beda0a7-8926-4ec5-a35a-337fe54d51a4">


### Comparison to standalone versions

<img width="742" alt="image" src="https://github.com/WebKit/Speedometer/assets/94007365/61f286af-7727-4139-9e6f-bf4024228003">

julienw commented on resources/todomvc/big-dom-generator/src/components/tree-area.jsx:
```diff
@@ -8,10 +9,10 @@ const TreeItem = (props) => {
     const { treeNode, currentDepth } = props;
 
     const isExpandableItem = treeNode.type === "expandableItem";
-    const treeViewItemIsOpen = isExpandableItem && currentDepth < MAX_VISIBLE_TREE_VIEW_ITEM_DEPTH ? "is-open" : "";
+    const treeViewItemIsOpen = isExpandableItem && (currentDepth !== MAX_VISIBLE_TREE_VIEW_ITEM_DEPTH || treeNode.isDisplayNone);
```
I don't understand this condition, it would be good to explain the logic of deciding if a node uses is-open or display-none or both.

lpardosixtosMs replied:
Added this comment "Every expandable TreeItem is open by default unless it is at the MAX_VISIBLE_TREE_VIEW_ITEM_DEPTH threshold. If it is marked as display none, set it as open so the `visibility: hidden` spectrum rule is removed.".

julienw commented on resources/todomvc/big-dom-generator/src/app.css:
```diff
@@ -86,6 +86,14 @@ Importing the CSS from the Spectrum CSS library.
+    display: none;
+}
+
+.spectrum-TreeView-item.is-open > .spectrum-TreeView {
+    visibility: unset;
+}
```
The impact of this CSS class is not clear to me. Which node does this apply to? Can you please add some more comments around that?

lpardosixtosMs replied:
Added this comment:
```
/*
The spectrum rule making the open tree view items visible is
.spectrum-TreeView-item.is-open > .spectrum-TreeView {
    visibility: visible;
}
This can cause descendants of closed tree view items to be visible.
We unset here to let it inherit the visibility of its parent.
*/
```

This is an example of the pathological behavior:
<img width="104" alt="image" src="https://github.com/WebKit/Speedometer/assets/94007365/bb7a7555-e184-41cb-9c3f-76ac7d3fb60d">

julienw replied:
I see, but why wasn't this happening before?

lpardosixtosMs replied:
> I see, but why wasn't this happening before?

Because before, our code was assuming that every tree item under a closed item is also closed, I changed it to only close the items at the threshold level. It makes more sense to me that the visibility of the children depends on the visibility of the parent instead of relying on its own `is-open` value.

julienw replied:
I see, can you write it as a comment too? I assume it's not just your code but spectrum in general assuming this.

lpardosixtosMs replied:
Done.

julienw reviewed (commented):
It's difficult to review without more comments in the code. I left a few comments to outline the parts where I'd need more explanations.
Thanks!

julienw commented on resources/todomvc/big-dom-generator/src/tree-generator.js:
```diff
@@ -3,6 +3,39 @@ import { DEFAULT_SEED_FOR_RANDOM_NUMBER_GENERATOR, MAX_GENERATED_DOM_DEPTH, MAX_
+    return node.subTreeWeight;
+};
+
+const markDisplayNoneNodes = (node, expandableItemWeight, nonExpandableItemWeight) => {
```
I see you added a comment in the next commit, sorry I missed it! Thanks!
 It could be expanded with more explanations about the logic, especially an example of the result.

lpardosixtosMs replied:
Done.

lpardosixtosMs commented on resources/todomvc/big-dom-generator/src/tree-generator.js:
```diff
@@ -3,6 +3,39 @@ import { DEFAULT_SEED_FOR_RANDOM_NUMBER_GENERATOR, MAX_GENERATED_DOM_DEPTH, MAX_
 
 const random = new LCG(DEFAULT_SEED_FOR_RANDOM_NUMBER_GENERATOR);
 
+const fillSubtreeWeights = (node, expandableItemWeight, nonExpandableItemWeight) => {
```
Done.

julienw commented on resources/todomvc/big-dom-generator/src/app.css:
```diff
@@ -86,6 +86,22 @@ Importing the CSS from the Spectrum CSS library.
+}
+
+/*
+The spectrum rule making the open tree view items visible is
+.spectrum-TreeView-item.is-open > .spectrum-TreeView {
+    visibility: visible;
+}
```
Can you add a few additional lines explaining that the normal TreeView has `visibility: hidden`?

Then explain that `unset` has the same effect as `visible` for the visible nodes, but doesn't have the side effect of showing descendants?

lpardosixtosMs replied:
Done.

julienw commented on resources/todomvc/big-dom-generator/src/components/tree-area.jsx:
```diff
@@ -8,10 +9,14 @@ const TreeItem = (props) => {
     const isExpandableItem = treeNode.type === "expandableItem";
-    const treeViewItemIsOpen = isExpandableItem && currentDepth < MAX_VISIBLE_TREE_VIEW_ITEM_DEPTH ? "is-open" : "";
+    /**
+     * Every expandable TreeItem is open by default unless it is at the MAX_VISIBLE_TREE_VIEW_ITEM_DEPTH threshold.
+     * If it is marked as display none, set it as open so the `visibility: hidden` spectrum rule is removed.
```
IMO this makes things a bit difficult to follow, what do you think about adding `visibility: unset` to the `.display-none` class?
BTW I think we also want to reset `height` to `auto` are we? Not sure this is necessary.

lpardosixtosMs replied:
This was just a cheap way of removing the `visibility: hidden` class from `.display-none` elements, doing `visibility: unset` inside the `.display-none` rule wouldn't work because the specificity of the original rule is higher. I think it should be fine to let `.display-none` elements have both `visibility: hidden` and `dispaly: none` properties, since `display: none` will take the elements out of the layout. Alternatively, I could use `!important` to unset them, but that seems even more artificial. WDYT?

julienw replied:
I think I'd use !important here, in this context, so that the 2 rules are in the same place. With some comment explaining why this is present.

lpardosixtosMs replied:
Done.

julienw reviewed (approved):
The comments help a lot, thank you

I still don't understand everything :D I left a few questions that I'd love to get answers about, but approving still.

flashdesignory reviewed (approved):
thanks @julienw for all the questions, that made it easier to somewhat follow along.

lpardosixtosMs commented on resources/todomvc/big-dom-generator/src/tree-generator.js:
```diff
@@ -1,8 +1,58 @@
+ * the target number of elements we want to have display none. Mark the
+ * nodes in the queue as display none.
+ * Consider the following example with the weights as displayed in the figure
+ * and a 10 as the target of display none elements. The iteration will stop
```
Done.

lpardosixtosMs commented:
@julienw I addressed all your feedback, I'll go ahead and merge the PR now. Please let me know if there are any more changes you'd like to see in a follow-up.
